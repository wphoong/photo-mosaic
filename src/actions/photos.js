import uuid from "uuid";

// ADD PHOTO

export const addPhoto = (
  {
    photoLink = "",
    title = "",
    createdAt = 0,
    description = ""
  } = {}) => ({
    type: "ADD_PHOTO",
    photo: {
      id: uuid(),
      photoLink,
      title,
      createdAt,
      description
    }
  }
);

export const editPhoto = (id, updates) => ({
  type: "EDIT_PHOTO",
  id,
  updates
});

export const removePhoto = ({id} = {}) => ({
  type: "REMOVE_PHOTO",
  id
});

// export const startAddPhoto = (photoData = {}) => {
//   return (dispatch, getState) => {
//     const uid = getState().auth.uid;
//     const {
//       photoLink = "",
//       title = "",
//       description = "",
//       createdAt = 0
//     } = photoData;
//     const photo = { photoLink, title, description, createdAt };
    
//     return dispatch(addPhoto({
//       id: ref.key,
//       ...photo 
//     }));
//   };
// };