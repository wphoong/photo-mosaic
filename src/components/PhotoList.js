import React from "react";
import { connect } from "react-redux";
import PhotoListItem from "./PhotoListItem.js";
import getVisiblePhotos from "../selectors/photos.js";

export const PhotoList = (props) => (
  <div>
    { props.photos.length !== 0 && <h3>Showing {props.photos.length} Photo(s)</h3>}
    { props.photos.length === 0 ? 
      (
        <div>
          <span>Showing 0 Photos</span>
        </div>
      ) : 
        <div className="columns is-multiline is-tablet">
          {
            props.photos.map((photo) => {
              return (
                <div className="column is-one-third">
                  <PhotoListItem key={photo.id} {...photo} />
                </div>
                );
            })
          }
        </div>
    }
  </div>
);

const mapStateToProps = (state) => ({
  photos: getVisiblePhotos(state.photos, state.filters)
});

export default connect(mapStateToProps)(PhotoList);