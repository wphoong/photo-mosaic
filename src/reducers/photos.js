const photosReducerDefaultState = [];

const photosReducer = (state = photosReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_PHOTO":
      return [
        ...state,
        action.photo
      ];
    default:
      return state;
  }
};

export default photosReducer;