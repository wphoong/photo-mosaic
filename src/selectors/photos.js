import moment from "moment";

const getVisiblePhotos = (photos, { text, sortBy, startDate, endDate}) => {
  return photos.filter((photo) => {
    const createdAtMoment = moment(photos.createdAt);
    const setTextMatch = photo.title.toLowerCase().includes(text.toLowerCase());
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;

    return setTextMatch && startDateMatch && endDateMatch;
  }).sort((a, b) => {
      if (sortBy === "descend") {
        return b.createdAt - a.createdAt;
      } else {
        return a.createdAt - b.createdAt;
      }
    }
  );
};

export default getVisiblePhotos;
