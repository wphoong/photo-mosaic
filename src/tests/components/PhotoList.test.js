import React from "react";
import { shallow } from "enzyme";
import { PhotoList } from "../../components/PhotoList.js";
import photos from "../fixtures/photos.js";

let wrapper;

beforeEach(() => {
  wrapper = shallow(<PhotoList photos={photos}/>);
});

test("should render photo list with provided photos", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should render 0 photos with no provided photos", () => {
  wrapper = shallow(<PhotoList photos={[]} />);
  expect(wrapper).toMatchSnapshot();
});