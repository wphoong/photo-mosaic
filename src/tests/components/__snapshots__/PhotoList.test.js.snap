// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 0 photos with no provided photos 1`] = `
<div>
  <div>
    <span>
      Showing 0 Photos
    </span>
  </div>
</div>
`;

exports[`should render photo list with provided photos 1`] = `
<div>
  <Connect(PhotoListItem)
    createdAt={0}
    description="descript1"
    id={1}
    key="1"
    photoLink="ez"
    title="test1"
  />
  <Connect(PhotoListItem)
    createdAt={-100}
    description="descript2"
    id={2}
    key="2"
    photoLink="ez2"
    title="test2"
  />
  <Connect(PhotoListItem)
    createdAt={1000}
    description="descript3"
    id={3}
    key="3"
    photoLink="ez3"
    title="test3"
  />
</div>
`;
